### PR TITLE
[MM-29733] Add e2e test case for configEditCmd

### PR DESCRIPTION
**Pull request by SezalAgrawal**
#### Summary

Add e2e test case for configEditCmd
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-29733
Fixes https://github.com/mattermost/mattermost-server/issues/15981

### CODE REVIEW

**Comment by SezalAgrawal** [comment]
@isacikgoz, I am getting `text file busy` error.  It is working on my local system. 
Not sure what could be the issue. I am creating a temp file and removing it later.

File: commands/config.go [hunk 1]
```diff
@@ -366,12 +366,14 @@ func configEditCmdF(c client.Client, _ *cobra.Command, _ []string) error {
 		editor = defaultEditor
 	}
 
+	fmt.Println(editor)
```
Comment (isacikgoz) [inline commands/config.go:369]:
```suggestion
```

File: commands/config.go [hunk 2]
```diff
@@ -366,12 +366,14 @@ func configEditCmdF(c client.Client, _ *cobra.Command, _ []string) error {
...
 	editorCmd := exec.Command(editor, file.Name())
 	editorCmd.Stdout = os.Stdout
 	editorCmd.Stdin = os.Stdin
 	editorCmd.Stderr = os.Stderr
 
 	if cmdErr := editorCmd.Run(); cmdErr != nil {
+		fmt.Println(cmdErr.Error())
```
Comment (isacikgoz) [inline commands/config.go:376]:
```suggestion
```

File: commands/config_e2e_test.go [hunk 3]
```diff
@@ -111,3 +115,60 @@ func (s *MmctlE2ETestSuite) TestConfigSetCmd() {
...
+		printer.Clean()
+
+		// check the value before editing
+		args := []string{"ServiceSettings.EnableSVGs"}
+		err := configGetCmdF(c, &cobra.Command{}, args)
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().False(*printer.GetLines()[0].(*bool))
+		s.Require().Len(printer.GetErrorLines(), 0)
+		printer.Clean()
```
Comment (isacikgoz) [inline commands/config_e2e_test.go:132]: I think we can safely use `s.th.App` for preconditions.
```golang
		// ensure the value before editing
		s.th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableSVGs = false })
```

File: commands/config_e2e_test.go [hunk 4]
```diff
@@ -111,3 +115,60 @@ func (s *MmctlE2ETestSuite) TestConfigSetCmd() {
...
+
+		file, err := ioutil.TempFile(os.TempDir(), "config_edit_*.sh")
+		s.Require().Nil(err)
+		defer func() {
+			file.Close()
+			os.Remove(file.Name())
+			resetCmd := &cobra.Command{}
+			resetCmd.Flags().Bool("confirm", true, "")
+			s.Require().Nil(configResetCmdF(c, resetCmd, []string{"ServiceSettings"}))
+		}()
```
Comment (isacikgoz) [inline commands/config_e2e_test.go:147]: If we use `s.th.App` utilities we may only need to close the file.
```suggestion
		defer file.Close()
```

File: commands/config_e2e_test.go [hunk 5]
```diff
@@ -111,3 +115,60 @@ func (s *MmctlE2ETestSuite) TestConfigSetCmd() {
...
+		_, err = file.Write([]byte(content))
+		s.Require().Nil(err)
+		editorCmd := exec.Command("chmod", "+x", file.Name()) //nolint:gosec
```
Comment (isacikgoz) [inline commands/config_e2e_test.go:150]: Is there a method where we don't need to bypass the security check?

File: commands/config_e2e_test.go [hunk 6]
```diff
@@ -111,3 +115,60 @@ func (s *MmctlE2ETestSuite) TestConfigSetCmd() {
...
+		os.Setenv("EDITOR", file.Name())
+
+		// check the value after editing
+		err = configEditCmdF(c, nil, nil)
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetErrorLines(), 0)
+		s.Require().Len(printer.GetLines(), 1)
+		config, ok := printer.GetLines()[0].(*model.Config)
+		s.Require().True(ok)
+		s.Require().True(*config.ServiceSettings.EnableSVGs)
```
Comment (isacikgoz) [inline commands/config_e2e_test.go:163]: Again we can use `s.th.App` methods to check if the value is actually changed or not.

File: commands/config_e2e_test.go [hunk 7]
```diff
@@ -111,3 +115,60 @@ func (s *MmctlE2ETestSuite) TestConfigSetCmd() {
...
+	})
+
+	s.Run("Edit config value without permissions", func() {
+		printer.Clean()
+
+		err := configEditCmdF(s.th.Client, nil, nil)
+		s.Require().NotNil(err)
```
Comment (isacikgoz) [inline commands/config_e2e_test.go:170]: We should check if we received the expected error here. Also you may want to utilize `s.Require().Error()` or `s.Require().EqualError()` method.

**Review by isacikgoz — CHANGES_REQUESTED** [review]
Thanks for the contribution @SezalAgrawal! I did my initial round of review. There are some opportunities to make test more simple :)

File: commands/config_e2e_test.go (hunk 5 above)
Comment (SezalAgrawal) [reply commands/config_e2e_test.go:141]: Will look into it. I wanted to make the shell script executable. I was getting `G204: Subprocess launched with function call as argument or cmd arguments` lint error, hence added this. Will try to see if I can find a better way to do this.

Comment (mgdelacroix) [reply commands/config_e2e_test.go:141]: Probably you can use [`os.Chmod`](https://golang.org/pkg/os/#Chmod) instead to add execution permissions, so the linter won't complain

**Review by mgdelacroix — APPROVED** [review]
Thanks for the contribution @SezalAgrawal! LGTM besides @isacikgoz's comments :+1:

**Review by lindy65 — APPROVED** [review]
Thanks @SezalAgrawal - looks good to me 👍 

//cc @mgdelacroix - here are the TM4J test codes for the E2E tests:
- Config - Edit MM-T3813
- Config - Edit : without permission MM-T3814

**Comment by emilyacook** [comment]
> Thanks @SezalAgrawal - looks good to me 👍
> 
> //cc @mgdelacroix - here are the TM4J test codes for the E2E tests:
> 
> * Config - Edit MM-T3813
> * Config - Edit : without permission MM-T3814

@mgdelacroix gentle ping, I think this is waiting on action from you?

**Comment by mgdelacroix** [comment]
@emilyacook thanks for the ping! It's waiting on the final pass from @isacikgoz, but he's OoO this week

In the meantime, @SezalAgrawal it looks like a couple of tests are failing in the CI:

```
        --- FAIL: TestMmctlE2ESuite/TestConfigEditCmd/Edit_a_key_in_config/SystemAdminClient (0.26s)
        --- FAIL: TestMmctlE2ESuite/TestConfigEditCmd/Edit_a_key_in_config/LocalClient (0.10s)
```

could you please take a look at them?

**Review by isacikgoz — APPROVED** [review]
Awesome @SezalAgrawal LGTM 🎉

**Pull request merged**